### ISSUE TITLE
Docker Crash when stopping container using ctrl+c

### ISSUE THREAD

**Issue by q-erty** [docker/for-mac#2765]
Hi all - 
I know this isn't the most useful bug report, but this was a pretty sudden crash during normal operations so I wanted to report it in case it's part of a larger pattern. Feel free to close if this isn't useful.

  - [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have submitted Diagnostics
  - [x] I have included the Diagnostics ID in this Issue

Diagnostic Id: 28A91478-3304-4C62-9000-539F69AE5D85/20180403-113943

### Expected behavior
Docker doesn't crash
### Actual behavior
Docker crashes
### Information
```
Docker for Mac: version: 18.03.0-ce-mac59 (dd2831d4b7421cf559a0881cc7a5fdebeb8c2b98)
macOS: version 10.12.6 (build: 16G1114)
logs: /tmp/28A91478-3304-4C62-9000-539F69AE5D85/20180403-113943.tar.gz
failure: com.docker.vpnkit is not running
[ERROR]  vpnkit
         Unexpected error ((Failure
  "Error connecting socket to 9p endpoint unix:/Users/ertyseidohl/Library/Containers/com.docker.docker/Data/s51: Unix.Unix_error(Unix.ECONNREFUSED, \"connect\", \"\")")) connecting to /Users/ertyseidohl/Library/Containers/com.docker.docker/Data/s51
         com.docker.vpnkit is not running
         vpnkit check failed with: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
[OK]     vmnetd
[OK]     dns
[ERROR]  driver.amd64-linux
         com.docker.driver.amd64-linux is not running
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     kubernetes
[OK]     files
[OK]     env
[OK]     virtualization kern.hv_support
[ERROR]  osxfs
         com.docker.osxfs is not running
[OK]     moby-console
[OK]     logs
[ERROR]  docker-cli
         Connection refused (ECONNREFUSED) connecting to /var/run/docker.sock: check if service is running
         Connection refused (ECONNREFUSED) connecting to /Users/ertyseidohl/Library/Containers/com.docker.docker/Data/s60: check if service is running
         docker ps failed
[OK]     disk
```

### Steps to reproduce the behavior

  1. Have a couple of containers running in TTY mode
  2. Stop one of the containers using ctrl+c

**Comment by q-erty** [comment]
Just crashed again. Here's what happned:

I started a docker container
I hit ctrl+c to bring down that docker container
I hit up+enter to restart that docker container (using docker-compose up)
The docker engine crashed

New diagnostic ID:
28A91478-3304-4C62-9000-539F69AE5D85

**Comment by OscarBarrett** [comment]
1 month later and I have had this happen once more in hundreds of times stopping containers with Ctrl-C. I've upgraded to `18.03.1-ce-mac65` since the last time.


```
^CGracefully stopping... (press Ctrl+C again to force)
Stopping myservice_nginx_1   ... done
Stopping myservice_php_1     ... done
Stopping myservice_mysql_1   ... error
Stopping myservice_node_1    ... done

ERROR: for myservice_mysql_1  b'Bad response from Docker engine'
```

Diagnostic ID C389B635-CB72-4209-BF80-541A3B0ADF61

```
Docker for Mac: version: 18.03.1-ce-mac65 (dd2831d4b7421cf559a0881cc7a5fdebeb8c2b98)
macOS: version 10.13.4 (build: 17E202)
logs: /tmp/C389B635-CB72-4209-BF80-541A3B0ADF61/20180530-103515.tar.gz
failure: com.docker.vpnkit is not running
[ERROR]  vpnkit
         Unexpected error ((Failure
  "Error connecting socket to 9p endpoint unix:/Users/oscarb/Library/Containers/com.docker.docker/Data/s51: Unix.Unix_error(Unix.ECONNREFUSED, \"connect\", \"\")")) connecting to /Users/oscarb/Library/Containers/com.docker.docker/Data/s51
         com.docker.vpnkit is not running
         vpnkit check failed with: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
[OK]     vmnetd
[OK]     dns
[ERROR]  driver.amd64-linux
         com.docker.driver.amd64-linux is not running
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     kubernetes
[OK]     files
[OK]     env
[OK]     virtualization kern.hv_support
[ERROR]  osxfs
         com.docker.osxfs is not running
[OK]     moby-console
[OK]     logs
[ERROR]  docker-cli
         Connection refused (ECONNREFUSED) connecting to /var/run/docker.sock: check if service is running
         Connection refused (ECONNREFUSED) connecting to /Users/oscarb/Library/Containers/com.docker.docker/Data/s60: check if service is running
         docker ps failed
[OK]     disk
```

**Comment by djs55** [comment]
In C389B635-CB72-4209-BF80-541A3B0ADF61/20180530-103515 I see:

```
2018-05-30 10:31:49.784127+0800  localhost com.docker.osxfs[96409]: Volume.stop b48dfc817ed1256f105d50e5c31706e43c51cde4ad9bc6cb88a479dde46ee861 (paths = [/Users/...:state=cached, /Users/.../php.ini:state=default])
2018-05-30 10:31:49.809626+0800  localhost com.docker.driver.amd64-linux[96411]: proxy >> POST /v1.24/containers/b48dfc817ed1256f105d50e5c31706e43c51cde4ad9bc6cb88a479dde
46ee861/wait
2018-05-30 10:31:49.809676+0800  localhost com.docker.driver.amd64-linux[96411]: Dial Socket vms/0/00000003.00000948
2018-05-30 10:31:49.809821+0800  localhost com.docker.driver.amd64-linux[96411]: proxy >> GET /v1.24/containers/b48dfc817ed1256f105d50e5c31706e43c51cde4ad9bc6cb88a479dde4
6ee861/json
2018-05-30 10:31:49.809839+0800  localhost com.docker.driver.amd64-linux[96411]: Dial Socket vms/0/00000003.00000948
2018-05-30 10:31:50.035177+0800  localhost com.docker.osxfs[96409]: While invalidating dcache of 18364 / functions_include.php: ENOTDIR
2018-05-30 10:31:50.062818+0800  localhost com.docker.driver.amd64-linux[96411]: Failed to remove container from osxfs: EOF
2018-05-30 10:31:50.089716+0800  localhost com.docker.supervisor[96408]: cannot read stdout from com.docker.osxfs: EOF
2018-05-30 10:31:50.091470+0800  localhost com.docker.supervisor[96408]: cannot read stderr from com.docker.osxfs: EOF
2018-05-30 10:31:50.091511+0800  localhost com.docker.supervisor[96408]: Reap com.docker.osxfs (pid: 96409): exit status 1
2018-05-30 10:31:50.096346+0800  localhost com.docker.supervisor[96408]: Stop 1 children with order 1: [com.docker.driver.amd64-linux (pid: 96411)]
2018-05-30 10:31:50.097768+0800  localhost com.docker.supervisor[96408]: Send signal terminated to com.docker.driver.amd64-linux (pid: 96411)
```

In C389B635-CB72-4209-BF80-541A3B0ADF61/20180427-111409
```
2018-04-27 11:13:47.238353+0800  localhost com.docker.osxfs[44621]: Volume.stop c829d6285913ecd2cc70222b5601080e31bfbd1dc53792487b03ffa392e10869 (paths = [/Users/...:state=cached, /Users/...:state=consistent])
2018-04-27 11:13:47.398880+0800  localhost com.docker.osxfs[44621]: While invalidating dcache of 2619 / components: ENOTDIR
2018-04-27 11:13:47.402782+0800  localhost com.docker.driver.amd64-linux[44623]: Failed to remove container from osxfs: EOF
2018-04-27 11:13:47.403366+0800  localhost com.docker.supervisor[44619]: cannot read stdout from com.docker.osxfs: EOF
2018-04-27 11:13:47.403416+0800  localhost com.docker.supervisor[44619]: cannot read stderr from com.docker.osxfs: EOF
2018-04-27 11:13:47.403486+0800  localhost com.docker.supervisor[44619]: Reap com.docker.osxfs (pid: 44621): exit status 1
2018-04-27 11:13:47.404142+0800  localhost com.docker.supervisor[44619]: Stop 1 children with order 1: [com.docker.driver.amd64-linux (pid: 44623)]
2018-04-27 11:13:47.404173+0800  localhost com.docker.supervisor[44619]: Send signal terminated to com.docker.driver.amd64-linux (pid: 44623)
2018-04-27 11:13:47.405172+0800  localhost com.docker.driver.amd64-linux[44623]: sending SIGTERM to com.docker.hyperkit pid: 44626
```

and in 28A91478-3304-4C62-9000-539F69AE5D85/20180403-113943
```
2018-04-03 11:39:30.816665-0400  localhost com.docker.osxfs[9198]: Volume.stop 6183bc44d17edcc42c522917eeb613377382639b433fd8629bc43ec8b01f51ca (paths = [/Users/...:state=cached])
2018-04-03 11:39:30.918175-0400  localhost com.docker.osxfs[9198]: While invalidating dcache of 24668 / web_modules: ENOTDIR
2018-04-03 11:39:30.937685-0400  localhost com.docker.driver.amd64-linux[9200]: Failed to remove container from osxfs: EOF
2018-04-03 11:39:30.951048-0400  localhost com.docker.supervisor[9197]: cannot read stderr from com.docker.osxfs: EOF
2018-04-03 11:39:30.951137-0400  localhost com.docker.supervisor[9197]: cannot read stdout from com.docker.osxfs: EOF
2018-04-03 11:39:30.951176-0400  localhost com.docker.supervisor[9197]: Reap com.docker.osxfs (pid: 9198): exit status 1
2018-04-03 11:39:30.955149-0400  localhost com.docker.supervisor[9197]: Stop 1 children with order 1: [com.docker.driver.amd64-linux (pid: 9200)]
2018-04-03 11:39:30.955475-0400  localhost com.docker.supervisor[9197]: Send signal terminated to com.docker.driver.amd64-linux (pid: 9200)
```

So I think the sequence is
- a volume is mounted with `:cached`
- the container is quit so the volume mount is deregistered
- as part of the de-registering the caching mode of the host path is set back to default (synchronous) so the caches in the VM must be flushed
- `osxfs` sends "please invalidate cache" messages to the VM
- `osxfs` receives `ENOTDIR` unexpectedly
- as this is considered a fatal should-never-happen error, `osxfs` quits
- all other processes are shutdown

I'll investigate this in more detail. It could well be that `ENOTDIR` is expected in some cases-- perhaps a directory was removed and a file created with the same name? Alternatively `osxfs` has become confused about whether this is a file or directory.

**Comment by q-erty** [comment]
To back up @djs55's point, we recently introduced `:cached` volumes and then started seeing this problem. Here's our docker-compose.yml:

```
---
version: "2"

services:
  api:
    build:
      context: .
      args:
        DEV_REQUIREMENTS: "TRUE"
    container_name: api
    networks:
        - default
        - (redacted)
    volumes:
      - .:/code:cached
    links:
      - mysql
      - redis
      - rabbitmq
      - ld-relay
    external_links:
      - (redacted)
    env_file:
      - ./env/env-lcl.txt
      - ./env/env-lcl-overrides.txt
    ports:
      - "8000:8000"
    command: [gunicorn, --config, gunicorn-dev.py, "q.wsgi:application"]
```

**Comment by djs55** [comment]
@q-erty thanks for the feedback!

Unfortunately I've not been able to trigger the problem myself. I'm also not sure yet whether the error should be considered harmless (the code already considers `ENOENT` to be harmless) or whether it is a sign of some underlying problem.

In the meantime I've updated `osxfs` to continue to log this error but not crash. I'm interested to find out how much this error occurs in practice and what affects it might have. For example it might cause some stale cache entries to persist in the VM which would prevent containers from picking up changes made on the host.

If you'd like to try an experimental build, here's one which shouldn't crash:
https://download-stage.docker.com/mac/bysha1/50a98f183efaf8f170ae3e2a8265dd6421cb2bc2/Docker.dmg

This is a development build which should only be used for experimentation (not production!) but if you have time to give it a shot, let me know how it goes. In particular watch out for stale cache entries appearing in the host, and have a look in your Mac's logs for the string "While invalidating dcache of". If you spot anything wrong, please upload a fresh set of diagnostics and I'll take a look.

**Comment by q-erty** [comment]
Thanks @djs55, I'll give it a try.

**Comment by RandalTeng** [comment]
emmmm, and this error have some solutions?
i think i got a same error with this,
how can i do? try with the development build of the link?

**Comment by rfay** [comment]
I tried the experimental build in https://github.com/docker/for-mac/issues/2765#issuecomment-393479468 and it seems to have completely solved the similar crash I have been able to cause repeatedly and predictably.  I have a situation that uses a docker volume and crashes on docker-compose stop

`Volume.stop fc0b843c51553726a4e7089aabdc45006c19c90c6ad719cc8f26e961dbf52a7d (paths = [/Users/rfay/workspace/d6git:state=cached, /Users/rfay/workspace/d6git/.ddev:state=default])
While invalidating inode of 318 (0 for -1): ENOTDIR
Failed to remove container from osxfs: EOF`

Normally it crashes at something between 1 and 40 iterations... and it's at 63 iterations right now, running strong. I'll leave it running all day.

I'd love to know what the change in this experimental build was and whether it is making it into docker 18.06 stable. We won't be able to release an important feature with this crash unless we can find some workaround or we enforce a docker upgrade.

**Comment by q-erty** [comment]
Just wanted to add that this issue seems to be resolved using the experimental build. Feel free to mark this as closed as soon as those changes make their way to master.

**Comment by rfay** [comment]
I find that *my* issue is resolved in docker-for-mac 18.06.0-ce-mac69 (26398), so maybe this made it in. I was regularly able to crash it in just a few iterations, and have now run my testing script through hundreds of iterations without a crash. I would like to hear if the changes mentioned here made it into 18.06.0-ce-mac69 (26398) though.

**Comment by djs55** [comment]
Sorry for the delay getting back to you. The fix should be in 18.05.0-ce-mac67 and later (including the current stable 18.06).

Thanks again for the report!
